auth0/resource_auth0_attack_protection.go

```diff
@@ -259,24 +263,61 @@ func readAttackProtection(d *schema.ResourceData, m interface{}) error {
 	return nil
 }
 
+func updateAttackProtection(d *schema.ResourceData, m interface{}) error {
+	api := m.(*management.Management)
+
+	if ipt := expandSuspiciousIPThrottling(d); ipt != nil {
+		if err := api.AttackProtection.UpdateSuspiciousIPThrottling(ipt); err != nil {
+			return err
+		}
+	}
+
+	if bfp := expandBruteForceProtection(d); bfp != nil {
+		if err := api.AttackProtection.UpdateBruteForceProtection(bfp); err != nil {
+			return err
+		}
+	}
+
+	if bpd := expandBreachedPasswordDetection(d); bpd != nil {
+		if err := api.AttackProtection.UpdateBreachedPasswordDetection(bpd); err != nil {
+			return err
+		}
+	}
+
+	return readAttackProtection(d, m)
+}
+
+func deleteAttackProtection(d *schema.ResourceData, m interface{}) error {
+	d.SetId("")
+	return nil
+}
+
 func flattenSuspiciousIPThrottling(ipt *management.SuspiciousIPThrottling) []interface{} {
 	m := make(map[string]interface{})
 	if ipt != nil {
 		m["enabled"] = ipt.Enabled
 		m["allowlist"] = ipt.AllowList
 		m["shields"] = ipt.Shields
-		m["pre_login"] = []interface{}{
-			map[string]int{
-				"max_attempts": ipt.Stage.PreLogin.GetMaxAttempts(),
-				"rate":         ipt.Stage.PreLogin.GetRate(),
-			},
-		}
-		m["pre_user_registration"] = []interface{}{
-			map[string]int{
-				"max_attempts": ipt.Stage.PreUserRegistration.GetMaxAttempts(),
-				"rate":         ipt.Stage.PreUserRegistration.GetRate(),
-			},
+
+		if ipt.Stage != nil {
+			if ipt.Stage.PreLogin != nil {
+				m["pre_login"] = []interface{}{
+					map[string]int{
+						"max_attempts": ipt.Stage.PreLogin.GetMaxAttempts(),
+						"rate":         ipt.Stage.PreLogin.GetRate(),
+					},
+				}
+			}
+			if ipt.Stage.PreUserRegistration != nil {
+				m["pre_user_registration"] = []interface{}{
+					map[string]int{
+						"max_attempts": ipt.Stage.PreUserRegistration.GetMaxAttempts(),
+						"rate":         ipt.Stage.PreUserRegistration.GetRate(),
+					},
+				}
+			}
```

Comment:
Similarly here, is this a precedent we want to set in the code? What good is this project if the SDK that it stands on contains such uncertainty? 

Another though I have here is if the Go SDK can smooth over some of these checks instead of pushing the responsibility to the TF provider; it'll just have to be implemented again on the Auth0 CLI end too.

Reply:
If we set a precedent with improving our error handling and having better defensive programming to avoid scenarios like segfaults, then yes. We do want to set a precedent for these kind of standards.

The philosophy of the go-auth0 SDK is to use pointers to attributes along with the `omitempty` json tag to remove parameters from getting deserialized. This is widespread across all the objects. The default value of a pointer is `nil`. If we access a nil object we get segfaults. We don't want segfaults. So we need better defensive programming around the philosophy that the go-auth0 SDK implements.